pyfr/plugins/fwh.py

```diff
@@ -0,0 +1,192 @@
+        d = np.linalg.norm(r_o, axis=-1)
+        r_o_hat = r_o / d[..., None]
+
+        m_r = r_o_hat @ self.qinf['M']
+
+        r_star_vec = r_o*np.hypot(gamma_inv, m_r)[..., None]
+        r_star_inv = 1 / np.linalg.norm(r_star_vec, axis=-1)
+
+        r_grad_fac = (np.einsum('ij,k->ijk', m_r, self.qinf['M']) +
+                      r_o_hat*gamma_inv**2)
+        r_snorm = r_o*r_star_inv[..., None]
+
+        r_star_tilde_vec = r_snorm*r_grad_fac
+        r_tilde_vec = (r_snorm - self.qinf['M'])*gamma**2
+
+        return r_tilde_vec, r_star_inv, r_star_tilde_vec
+
+    def fwh_solve(self, intg):
+        o_vals = np.zeros(self.nobvs)
+        ci = 1 / self.qinf['c']
+
+        # Querry dt_soln to prevent MPI deadlock
+        dt_soln = intg.dt_soln
+
+        # Accumulate FWH contribution from each surface part
+        for (etype, fidx), p in self.surf.items():
+            eidx, m0, n, a, r_tilde_v, r_star_inv, r_star_tilde_v = p
+
+            soln = intg.soln[self.emap[etype]][..., eidx]
+            soln_t = dt_soln[self.emap[etype]][..., eidx]
+
+            s = m0 @ soln.transpose(1, 0, 2)
+            s_t = m0 @ soln_t.transpose(1, 0, 2)
+
+            pris = self.elementscls.con_to_pri(s, self.cfg)
+
+            pris = np.reshape(pris, (self.nvars, -1))
+
+            p = pris[self._pidx] - self.qinf['p']
+            u = pris[self._vidx] - self.uinf
+            d_inf = self.qinf['rho']
+            d_tot = d_inf + p*ci**2
+
+            mom = d_tot*(u + self.uinf)
+            drift = -d_inf*self.uinf
+
+            # Time derivatives
+            pris_t = self.elementscls.diff_con_to_pri(s, s_t, self.cfg)
+            pris_t = np.reshape(pris_t, (self.nvars, -1))
+
+            u_t = pris_t[self._vidx]
+            p_t = pris_t[self._pidx]
+
+            d_tot_t = p_t / self.qinf['c']**2
+            mom_t = d_tot_t*(u + self.uinf) + d_tot*u_t
+            mom_t_n = np.sum(n*mom_t, axis=0, keepdims=True)
+
+            # Monopole
+            q = np.sum(n*(mom + drift), axis=0, keepdims=True).T
+            q_t = mom_t_n.T
+            I = (1 - self.Minf)*q_t*r_star_inv
+            I -= (r_star_tilde_v @ self.uinf.flatten())*q*r_star_inv**2
+
+            # Dipole
+            mom_n = np.sum(n*mom, axis=0, keepdims=True)
+            F = mom_n*u + (p - self.qinf['p'])*n
+            F_t = mom_t_n*u + mom_n*u_t + p_t*n
+            I += ci*r_star_inv*np.einsum('ki,ijk->ij', F_t, r_tilde_v)
+            I += np.einsum('ki,ijk->ij', F, r_star_tilde_v)*r_star_inv**2
+
+            # Quadrature and accumulate
+            o_vals += a @ I
+
+        return o_vals / (4*np.pi)
+
+    def __call__(self, intg):
+        if intg.tcurr < self.tstart:
+            return
+
+        comm, rank, root = get_comm_rank_root()
+
+        if intg.tcurr - self.dt >= self.t_last - self.tol:
+            self.t_last = intg.tcurr
+
+            o_vals = self.fwh_solve(intg)
+            comm.Barrier()
```

Comment:
Why is this needed?

Reply:
Reduce includes a barrier.

Reply:
It will be an MPI library specific thing, but the timeout is certainly not part of the MPI spec.  Also, we use a similar reduce call in several other plugins without issue.

Reply:
Just following up here.